extensions/barrage/src/main/java/io/deephaven/extensions/barrage/util/BarrageStreamReader.java

```diff
@@ -219,30 +238,77 @@ public BarrageMessage safelyParseFrom(final StreamReaderOptions options,
                     }
                     final TLongIterator bufferInfoIter = bufferInfo.iterator();
 
-                    final boolean isAddBatch = numAddBatchesRemaining > 0;
-                    if (isAddBatch) {
-                        --numAddBatchesRemaining;
-                    } else {
-                        --numModBatchesRemaining;
-                    }
+                    // add and mod rows are never combined in a batch. all added rows must be received before the first
+                    // mod rows will be received
 
-                    if (isAddBatch) {
-                        final int numRowsTotal = msg.rowsIncluded.intSize("BarrageStreamReader");
+                    if (numAddRowsRead < numAddRowsTotal || (numAddRowsTotal == 0 && numModRowsTotal == 0)) {
                         for (int ci = 0; ci < msg.addColumnData.length; ++ci) {
-                            msg.addColumnData[ci].data = ChunkInputStreamGenerator.extractChunkFromInputStream(options,
-                                    columnChunkTypes[ci], columnTypes[ci], componentTypes[ci], fieldNodeIter,
-                                    bufferInfoIter, ois, (WritableChunk<Values>) msg.addColumnData[ci].data,
-                                    numAddRowsRead, numRowsTotal);
+                            final BarrageMessage.AddColumnData acd = msg.addColumnData[ci];
+
+                            int lastChunkIndex = acd.data.size() - 1;
+
+                            // need to add the batch row data to the column chunks
+                            WritableChunk<Values> chunk = (WritableChunk<Values>) acd.data.get(lastChunkIndex);
+                            int chunkSize = chunk.size();
```

Comment:
So, for the snapshot case where you've added `null` won't this NPE? 

It seems incorrect now for `extractChunkFromInputStream` to have the feature to return/replace the provided chunk. As in, it seems like we could make it an error if a large enough chunk is not provided. You can fix the NPE for this PR, and change the extractChunk interface in a follow up?

Reply:
Moved to #2365